api/src/main/java/org/apache/iceberg/transforms/Bucket.java

```diff
@@ -124,19 +127,43 @@ public String toString() {
     }
   }
 
+  @Override
+  public UnboundPredicate<Integer> project(String name, BoundSetPredicate<T> predicate) {
```

Comment:
This PR made me realize that adding `BoundSetPredicate` not only required adding a new method to the `ExpressionVisitor`, but also required new methods here in the `Transform` API. Because of this, I think a better option is to refactor to make `BoundSetPredicate` extend `BoundPredicate`. I've implemented that in #650; please take a look at that PR since I think this one should depend on those changes.